Add 2.7 blog post (#12966)
prettier/prettier: website/blog/2022-06-11-2.7.0.md

## Highlights

### TypeScript

#### Support TypeScript 4.7 ([#12896](https://github.com/prettier/prettier/pull/12896), [#12897](https://github.com/prettier/prettier/pull/12897), [#12898](https://github.com/prettier/prettier/pull/12898), [#12900](https://github.com/prettier/prettier/pull/12900), [#12921](https://github.com/prettier/prettier/pull/12921), [#12924](https://github.com/prettier/prettier/pull/12924), [#12959](https://github.com/prettier/prettier/pull/12959) by [@sosukesuzuki](https://github.com/sosukesuzuki))

Support [TypeScript 4.7](https://devblogs.microsoft.com/typescript/announcing-typescript-4-7) new features!

##### [Instantiation Expressions](https://devblogs.microsoft.com/typescript/announcing-typescript-4-7/#instantiation-expressions)

<!-- prettier-ignore -->
```ts
interface Box<T> {
  value: T;
}
function makeBox<T>(value: T) {
  return { value };
}
const makeHammerBox = makeBox<Hammer>;
const makeWrenchBox = makeBox<Wrench>;
```

> ##### [Optional Variance](https://devblogs.microsoft.com/typescript/announcing-typescript-4-7/#optional-variance-annotations-for-type-parameters)

Reviewer: ```suggestion
##### [Optional Variance Annotations](https://devblogs.microsoft.com/typescript/announcing-typescript-4-7/#optional-variance-annotations-for-type-parameters)
```

<!-- prettier-ignore -->
```ts
interface Animal {
  animalStuff: any;
}
interface Dog extends Animal {
  dogStuff: any;
}
type Getter<out T> = () => T;
type Setter<in T> = (value: T) => void;
```

##### [`extends` constraints for `infer`](https://devblogs.microsoft.com/typescript/announcing-typescript-4-7/#extends-constraints-on-infer-type-variables)

<!-- prettier-ignore -->
```ts
type FirstString<T> = T extends [infer S extends string, ...unknown[]]
  ? S
  : never;
```

### CLI

> #### Add `--cache` and `--cache-strategy` CLI option ([#12800](https://github.com/prettier/prettier/pull/12800) by [@sosukesuzuki](https://github.com/sosukesuzuki))

Reviewer: ```suggestion
#### Add `--cache` and `--cache-strategy` CLI options ([#12800](https://github.com/prettier/prettier/pull/12800) by [@sosukesuzuki](https://github.com/sosukesuzuki))
```

  "does something really long and complicated so I have to write a very long name for the test",
  () => {}
);

// Prettier 2.7
test.step("does something really long and complicated so I have to write a very long name for the test", () => {});

test.describe("does something really long and complicated so I have to write a very long name for the test", () => {});

test.describe
  .only("does something really long and complicated so I have to write a very long name for the test", () => {});

test.describe
  .parallel("does something really long and complicated so I have to write a very long name for the test", () => {});

test.describe.parallel
  .only("does something really long and complicated so I have to write a very long name for the testThis is a very", () => {});

test.describe
  .serial("does something really long and complicated so I have to write a very long name for the test", () => {});

test.describe.serial
  .only("does something really long and complicated so I have to write a very long name for the test", () => {});
```

#### Fix comment formats in ([#12860](https://github.com/prettier/prettier/pull/12860) by [@HosokawaR](https://github.com/HosokawaR))

> This change fixes comments format in `exports` to algin with the comment format in `import`

Reviewer: ```suggestion
This change fixes the comment format in `exports` to align with the comment format in `import`.
```

> Although this change does not affect the comments format in `import`, follows change log contains examples of comments in `import` for reference.

Reviewer: ```suggestion
Although this change does not affect the comments format in `import`, the following changeset contains examples of comments in `import` for reference.
```

(Skeptical about either "change log" or "changeset" being the right word, but nothing better on my mind)

<!-- prettier-ignore -->
```jsx
// Input
export {
  foo,
  
  bar as  // comment
		 baz,
}

import {
  foo,
  
  bar as  // comment
		 baz,
} from 'foo'

// Prettier 2.6
export {
  foo,
  bar as baz, // comment
};

import {
  foo,
  // comment
  bar as baz,
} from "foo";

// Prettier 2.7
export {
  foo,
  // comment
  bar as baz,
};

import {
  foo,
  // comment
  bar as baz,
} from "foo";

```

### TypeScript

#### Print `as` instead of `:` for `babel-ts` parser ([#12706](https://github.com/prettier/prettier/pull/12706) by [@HosokawaR](https://github.com/HosokawaR))

<!-- prettier-ignore -->
```tsx
// Input
[x as any] = x;

// Prettier 2.6
[x: any] = x;

// Prettier 2.7
[x as any] = x;
```

#### Fix formatting for typescript Enum with computed members ([#12930](https://github.com/prettier/prettier/pull/12930) by [@HosokawaR](https://github.com/HosokawaR))

<!-- prettier-ignore -->
```tsx
// Input
enum A {
  [i++],
}

// Prettier 2.6
enum A {
  i++,
}

// Prettier 2.7
enum A {
  [i++],
}
```

### HTML

#### Support Speculation Rules API formatting in HTML ([#12882](https://github.com/prettier/prettier/pull/12882) by [@sosukesuzuki](https://github.com/sosukesuzuki))

> Please read https://web.dev/speculative-prerendering/ for about Speculation Rules API.

Reviewer: ```suggestion
Please read https://web.dev/speculative-prerendering/ for more information about the Speculation Rules API.
```

  name: String
}

""""""
type Person {
  name: String
}

# Prettier 2.6.2
"""
 Customer
"""
type Person {
  name: String
}

"""

"""
type Person {
  name: String
}

# Prettier 2.6.2 (second format)
"""
Customer
"""
type Person {
  name: String
}

"""

"""
type Person {
  name: String
}

# Prettier 2.7
"""
Customer
"""
type Person {
  name: String
}

"""
"""
type Person {
  name: String
}
```

### CLI

#### Print the number of files need change ([#12561](https://github.com/prettier/prettier/pull/12561) by [@Harry-Hopkinson](https://github.com/Harry-Hopkinson))

```console
# Prettier 2.6
$ prettier src --check
Checking formatting...
[warn] src/fileA.js
[warn] Code style issues found in the above file(s). Forgot to run Prettier?

# Prettier 2.7
$ prettier src --check
Checking formatting...
[warn] src/fileA.js
[warn] src/fileB.js
[warn] Code style issues found in 2 files. Forgot to run Prettier?

$ prettier src --check
Checking formatting...
[warn] src/fileA.js
[warn] Code style issues found in the above file. Forgot to run Prettier?
```

#### Infer parser for `.importmap` files ([#12603](https://github.com/prettier/prettier/pull/12603) by [@fisker](https://github.com/fisker))

> Format [`.importmap`](https://github.com/WICG/import-maps#import-map-processing) files as a JSON files.

Reviewer 2: ```diff
Format [`.importmap`](https://github.com/WICG/import-maps#import-map-processing) files as JSON files.
```
